### PR TITLE
make IngredientNBT constructor public

### PR DESCRIPTION
**Pull request by AEnterprise**
This allows mods to create an IngredientNBT from code for recipes that are dynamicaly generated but have ingredients that are NBT sensitive.

In this case it would be for dynamicaly generating recipes for crafting gates for all combinations of registered materials and logic modifiers

### CODE REVIEW

**Comment by Xalcon** [comment]
I would like to see this merged as well. It would make custom ingredient factories (workaround for #4113) simpler, because at the moment, people need to extend IngredientNBT just to get access to the constructor.

**Comment by LexManos** [comment]
No, and I will explicitly state this.
This is to make code recipes harder. You're supposed to use JSON, use it.
This is another case of "I wanna use it in code because I wanna!"
Those type of arguments are not welcome here. If you can articulate a real reason beyond you don't like json then I may consider it but it was a conscious decision to NOT expose this in code.

**Comment by trentv4** [comment]
Lex, what is the recommended system for implementing dynamic NBT recipes? JSON files are impractical in some situations where ingredients are determined at runtime.

**Comment by LexManos** [comment]
Json. Be more specific 'dynamic' means NOTHING. "JSON files are impractical in some situation" NAME THOSE SITUATIONS. Seriously. You just pulled a "I want it because i want it!" argument.
We will not be adding things to allow you to bypass future facing features because you can't articulate what is wrong with them and arnt willing to work on improving them.

**Comment by shadowfacts** [comment]
The original post gave a specific situation:

> In this case it would be for dynamicaly generating recipes for crafting gates for all combinations of registered materials and logic modifiers

**Comment by Madsthunder** [comment]
I think it would be better to just implement a dynamic recipe system for json files.

**Comment by LexManos** [comment]
Ya, you don't have a crapload of recipes. You have one. 
'Add this material to this item'
No need for all the random permutations or anything. 
Your Ingredient wouldn't be a IngredientNBT it would be an IngredientMaterial, or IngredientAllMaterials.

My point is you guys are stuck on the concept of 'I MUST create 10,000 recipes because PERMUTATIONS!!!' That is not how you should be thinking. In reality you're only creating one recipe and copy pasting it a crapload of times. Address the json as that.

**Comment by AEnterprise** [comment]
alright i do have to admit that in this specific PR case a big part of not moving to json is the amount of work needed for it

but as for what is wrong with the json system for us as developers atm (correct me if i'm wrong but this is based on me talking to other devs mostly) is the overhead of having to maintain those 10,000 files that are all basicly the same except for minor differences

especially for dynamic recipes where you need some properties of an input item to determine the properties of the output moving to json is alot of work for not realy any gain to us atm

for those recipes you need a custom recipe parser to create and return a IRecipe together with custom ingredients (not always but most of the time). This means alot of extra code just to be able to build up recipes by reading a few bits out of a json file. and with the current system there basicly is NO real difference between a recipe that a mod assembled in code using data from a json (that the modder defined himself) or a recipe that is completely build up in code, for simple recipes where there is no dyamic part i'm perfectly fine with JSON

But for dynamic ones where the output changes sligtly with the input (pipes for example the glass color determines the color of the pipe glass, it would be alot of work to move all of those to JSON. Yes it's copy paste, doing them all by hand means hours and hours of work, using factories it'll be less but still a fair bit of work. The only way an automated converter does it would be to dump all permutations in files, works but need to write the code to build up the recipes first (and once you have the code why export to json? you might as well parse it to a recipe instead and register it anyways). Having those 2000 recipes from the converter means you need to keep the code if you later want to change one of the core ingredients or the recipe layout (unless you want to correct 2000 files by hand with the chanse of missing one)

It could also be done pipe by pipe to generate jsons to use constants, easier to change some of the components in the recipe for another but recipe layouts changes our output counts mean you again have 2000 files to fix by hand or you need to keep the code again to be able to re-run it later. Also it's pipe by pipe then so again this take time and effort

looking at both ways it would take an hour or more to get it all working properly if not more

If you compare that workload and time investement to writing a few simple loops that assemble and register the recipe (and we already had those or you would need to write most if not all of those anyways during the previous mentioned methods) the effort required to move them all and maintain in case you want to change is just not realy worth the time required to do it for most modders.

Simple recipes that have no variations depending on changing one of the inputs or similar are much easier and no code to help write the jsons would be needed so in that case just going straight for json makes sense and is worth it


So yes, this whole argument seems to always return to or can be boiled down to an argument of investing time into doing it. While the offical stance of the forge team is that that is not an valid argument to most modders it is a completely valid and critical argument even. We are all hobbyist, we make mods because we love making mods for the comunity but we all have lives and jobs and other things to do. We are a community of people who make things for others for FUN, we want to have fun and provide fun, not having to grind our way through writing 2000 boring json files while we could be spending that time working on new features, new content, more bugfixes, ...

TLDR: yes it boils down to an argument just being willing to spend the time writing all the files, but i beleive (and i am conviced alot of modders agree with me here) that time and effort are very important aspects when developing something. Fr me a modding framework is ment to give modders the tool to be able to do the things they want in a good, clean, efficient and fast way. Not making it harder

**Comment by AEnterprise** [comment]
also after some discussion with others: you say the argument of 'code because code' is an invalid argument, yet the only counter argument seems to be 'json because json'

the code and json systems should be able to coëxist next to each other and complement each other, i do not see why the the code system should be 'handicaped' or made harder to use just because the json system exists.

yes you can do anything you did with the code system in the json system, but that works both ways: anything you do in the json system can be done in the code system. before json there where mods who even defined their recipes in xml for machines (could have been json just as easely).

So why can't we just allow people to use both side by side, picking the system that best suits their needs, instead of limiting one just so they would use the one you like most?

**Comment by Xalcon** [comment]
> also after some discussion with others: you say the argument of 'code because code' is an invalid argument, yet the only counter argument seems to be 'json because json'

The reason is - as lex said - future proofing. In the future, it is highly likely that recipes will be synced from the server to the client. To do this, the recipes need to be serializable or even better: already in a serialized format. If you register recipes in code, those recipes would need to supply a way to be serialized... hence the reason why json is a thing in the first place.

**Comment by AEnterprise** [comment]
even with json that would not be possible, it would still only work if the code for the factories is also available on the client, any ingredient factory that would be ment for checking specific NBT tags while still ignoring others (to give a specific example: checking the gate type but ignoring the slot count) will not work on the client unless the code for that specific ingredient would also be available on the client.

The server would have it and match the recipe on the server but the client would still be as much in the dark as when the recipe was fully defined in code

**Comment by Xalcon** [comment]
ingredient factories are part of the Mod code and are therefore bound to a mod version. If versions dont match or are incompatible, it wont work. This is nothing new.

**Comment by AEnterprise** [comment]
sorry seems i misunderstood your point before

servers will indeed sync recipes to the client in the future, there extra recipes will HAVE to be json most likely but that is a design limitation with how datapacks work (or will work as they are not released yet)

the only recipes that would have to be synced to the client would be aditional recipes, recipes added by pack/world makers, ment to override or suplement already existing recipes.

since all mods on the server also have to be on the client i do not see why recipes added by those mods would need to be synced to the client again, they would already exist. Forge would only have to send those additional recipes since the modded recipes would already be there.

There simply is no reason i can see to have mod recipes be serialized when they would not have to be synced to the client (also with any major modpack this would be a terrible idea, imagine having to send 10000+ recipes to the client each time it connects)

**Comment by LexManos** [comment]
There is nothing wrong with sending 10,000 recipes to the client. At the end of the day this would be trivial. And my argument isn't 'json for jsons sake' like yours is. Its Json for future compatibility, simplicity, sane management, and configurability.
Yes, the factory for a recipe would have to be on both sides, however having a handful of factories in sync is FAR easier than assuming ALL of the recipes are in sync.
The point is to start making your recipes generic. As I have stated many, many times, you don't need 10,000 recipes that do the EXACT same thing just with one flag switched based on input.

The entire point of this is that JSON is the future, and SHOULD be what you're working towards. It has A LOT of potential benefits over code once the system has matured. I would like to see the system mature. "Gimmie da old stuff cuz change is scary!" arguments do not help things mature. So I will not tolerate them. This PR literally saves you 1 line of code. At the expense of encouraging modders like you to sit in your shelters and not even consider any other way. 

Forge's primary job is NOT to make modders lives 'easier', it is to make Minecraft the best engine it can and to help as many mods as possible be compatible with each other. Sometimes that means telling you modders No, you're doing it wrong!

Anyways, i'm going to lock this as it appears to be impossible to even make you consider changing your mind. I am open to discussions on how to make the JSON/Configurable recipe system better and easier to use. I am NOT open to people scared of change whining to get the old system back.
